fhir-config/src/main/java/com/ibm/fhir/config/ConfigurationService.java

```diff
@@ -53,7 +58,9 @@ public static PropertyGroup loadConfiguration(String filename) throws Exception
      *            an InputStream to the input JSON file
      */
     public static PropertyGroup loadConfiguration(InputStream is) throws Exception {
-        try (JsonReader reader = JSON_READER_FACTORY.createReader(is)) {
+        String templatedJson = IOUtils.toString(is, StandardCharsets.UTF_8);
+        String resolvedJson = StringSubstitutor.replace(templatedJson, System.getenv());
```

Comment:
The use of such a prefix is of course possible and might be helpful to distinct environment variables targeted for the FHIR server from the rest - yet I am just not convinced whether this should be mandatory. Rather I believe that such hard-coded prefix could possibly restrict the user from fully following and complying to his/her own, local naming conventions for variables. Also OpenLiberty itself does not stipulate any prefix either for (environment) variables used in its XML configuration files, if I see that correctly. What do you think?

Reply:
I think it's best to prefix.  While it constrains the environment, it doesn't permit needless introspection into the environment variables that may be beyond the scope of the application.  

Maybe it's best to have a second reviewer comment? @lmsurpre thoughts?

Reply:
I am sorry but I must admit that I do not understand what you mean with "permit needless introspection into the environment variables that may be beyond the scope of the application". The substitution of the variables within the JSON configuration files happens directly "on-the-fly" when loading those files within `ConfigurationService.loadConfiguration(InputStream is)` and hence the actual application, i.e. the FHIR server, does not see the variables themselves but only their (final) values and thus does not have to perform any further introspection.

Reply:
Right, I understand that, however this code also is permitted to run in a non-docker environment.  An attacker could conceivably gain access to more info than required, if not prefixed.

Reply:
I am sorry again for my ignorance but I am do not get the point how a prefix `FHIR.` could help to have an attacker gain access to less information. For example, using `postgres.password` as plain environment variable is definitely "suboptimal" security-wise but I do not understand how `FHIR.postgres.password` would ameliorates things. Can you perhaps give me an example/use case with a "malicious" variable so that I can better get what you mean? That would be great!

Reply:
So it's a reference to a secondary attack, where you inject a new configuration which may have access to an unexpected or unscoped environment variable. 

Also, a good lesson from CERT https://wiki.sei.cmu.edu/confluence/display/c/ENV02-C.+Beware+of+multiple+environment+variables+with+the+same+effective+name and more specific to java https://wiki.sei.cmu.edu/confluence/display/java/ENV02-J.+Do+not+trust+the+values+of+environment+variables

I like the thought of scoping it as it is an intent to use versus a side effect. 

PRO
no changes, user defines completely

CON
indirect re-use of environment variables (potential overrides)
potential conflicts with other Environment variables

Reply:
Thanks a lot for providing me with the information and links. Yet I do not understand how the prefix could solve the linked issues: Following the first link, this only makes the case for having clearly distinct environment variable names without relying on case difference - so if you have `FHIR.some.variable` set to a certain value then a malicious attacker could also analogously change the prefix casing to "redefine" that value by simply setting e.g. `Fhir.some.variable`, right? The second link makes the case for using pre-defined Java system *properties* and does not mention the usage of prefixes at all - and so at the end of the day it boils down to simply not use freely defined environment variables at all - no matter if prefixed or not, right? So in the end the coping by introducing the prefix seems to be actually only a pure naming convention but would have no restrictive effect or similar - or am I still missing anything possibly?

Reply:
I'm still leaning towards prefixing.  I'll leave it to John and Lee to make the final call.